appsmithorg/appsmith
# fix: unresponsive TextWidget when bound to large data

## Description

Cause - When TextWidget is bound to large data (JSON / html) through an api, then the screen becomes unresponsive due to creation to large number of DOM nodes.

Fix - 
1. Enable ```shouldTruncate``` by default.
2. In TextComponent, if truncation is enabled; then calculate the maximum length of text that can be displayed in the associated text box ( = widget area / rough area covered by a single character). Store the substring as a state variable ```displayText```.

Fixes #10771


## Type of change


- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
1. Pull the branch.
2. Make an api pointing to https://restcountries.com/v3.1/all
3. Bind it to a text widget.
4. Upon refresh the TextWidget should render the truncated text without becoming unresponsive.
5. Click on ellipses icon to launch the modal component. Observe that the complete text is rendered in the modal.


## Checklist:







## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>


    // Code coverage diff between base branch:release and head branch: fix/10771 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.74 **(0.03)** | 37.13 **(0.05)** | 35.8 **(0.02)** | 56.09 **(0.02)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :green_circle: | app/client/src/widgets/TextWidget/component/index.tsx | 86.67 **(13.59)** | 50.23 **(5.2)** | 65.38 **(7.69)** | 85.71 **(13.34)**</details>
